Framework/Source/Graphics/Program/ProgramReflection.cpp

```diff
@@ -0,0 +1,1190 @@
+        if ((kind == TypeReflection::Kind::Matrix) || (kind == TypeReflection::Kind::Vector) || (kind == TypeReflection::Kind::Scalar))
+        {
+            storeShaderVariable(path, category, name, varMap, pVarMapBySemantic, 0, 0);
+        }
+        else if (kind == TypeReflection::Kind::Array)
+        {
+            auto arrayKind = pTypeLayout->getElementTypeLayout()->getKind();
+            assert((arrayKind == TypeReflection::Kind::Matrix) || (arrayKind == TypeReflection::Kind::Vector) || (arrayKind == TypeReflection::Kind::Scalar));
+            uint32_t arraySize = (uint32_t)pTypeLayout->getTotalArrayElementCount();
+            uint32_t arrayStride = (uint32_t)pTypeLayout->getElementStride(SLANG_PARAMETER_CATEGORY_UNIFORM);
+            storeShaderVariable(path, category, name, varMap, pVarMapBySemantic, arraySize, arrayStride);
+        }
+        else if (kind == TypeReflection::Kind::Struct)
+        {
+            for (uint32_t f = 0; f < pTypeLayout->getFieldCount(); f++)
+            {
+                ReflectionPath newPath;
+                newPath.pVar = pTypeLayout->getFieldByIndex(f);
+                newPath.pParent = &path;
+                newPath.childIndex = f;
+                std::string memberName = name + '.' + newPath.pVar->getName();
+                reflectVaryingParameter(newPath, memberName, category, varMap, pVarMapBySemantic);
+            }
+        }
+        else
+        {
+            should_not_get_here();
+        }
+    }
+
+    static void reflectShaderIO(EntryPointReflection* pEntryPoint, SlangParameterCategory category, ProgramReflection::VariableMap& varMap, ProgramReflection::VariableMap* pVarMapBySemantic = nullptr)
+    {
+        uint32_t entryPointParamCount = pEntryPoint->getParameterCount();
+        for (uint32_t pp = 0; pp < entryPointParamCount; ++pp)
+        {
+            ReflectionPath path;
+            path.pVar = pEntryPoint->getParameterByIndex(pp);
+            reflectVaryingParameter(path, path.pVar->getName(), category, varMap, pVarMapBySemantic);
+        }
+    }
+
+    ProgramReflection::SharedPtr ProgramReflection::create(slang::ShaderReflection* pSlangReflector, std::string& log)
+    {
+        return SharedPtr(new ProgramReflection(pSlangReflector, log));
+    }
+
+    ProgramReflection::ProgramReflection(slang::ShaderReflection* pSlangReflector, std::string& log)
+    {
+        ParameterBlockReflection::SharedPtr pGlobalBlock = ParameterBlockReflection::create("");
+        for (uint32_t i = 0; i < pSlangReflector->getParameterCount(); i++)
+        {
+            VariableLayoutReflection* pSlangLayout = pSlangReflector->getParameterByIndex(i);
+            ReflectionVar::SharedPtr pVar = reflectTopLevelVariable(pSlangLayout);
+
+            // In GLSL, the varying (in/out) variables are reflected as globals. Ignore them, we will reflect them later
+            if (pVar->getType()->unwrapArray()->asResourceType() == nullptr) continue;
```

Comment:
This logic seems a bit off; what if the user has a global-scope non-varying parameter that *doesn't* have a resource type? Maybe somebody writes:

```hlsl
struct Foo { Texture2D t; SamplerState s; }
Foo foo;
```

There `foo` is a valid HLSL/Slang shader parameter that doesn't have a resource type (although it does contain resources).